extensions/ql-vscode/src/debugger/debug-session.ts

```diff
@@ -0,0 +1,324 @@
+    response.body.supportsStepBack = false;
+    response.body.supportsStepInTargetsRequest = false;
+    response.body.supportsRestartFrame = false;
+    response.body.supportsGotoTargetsRequest = false;
+
+    this.sendResponse(response);
+
+    this.sendEvent(new InitializedEvent());
+  }
+
+  protected configurationDoneRequest(
+    response: DebugProtocol.ConfigurationDoneResponse,
+    args: DebugProtocol.ConfigurationDoneArguments,
+    request?: DebugProtocol.Request,
+  ): void {
+    super.configurationDoneRequest(response, args, request);
+  }
+
+  protected disconnectRequest(
+    response: DebugProtocol.DisconnectResponse,
+    _args: DebugProtocol.DisconnectArguments,
+    _request?: DebugProtocol.Request,
+  ): void {
+    response.body = response.body ?? {};
+    // Neither of the args (`terminateDebuggee` and `restart`) matter for CodeQL.
+
+    this.sendResponse(response);
+  }
+
+  protected launchRequest(
+    response: DebugProtocol.LaunchResponse,
+    args: CodeQLDebugProtocol.LaunchRequestArguments,
+    _request?: DebugProtocol.Request,
+  ): void {
+    void this.launch(response, args); //TODO: Cancelation?
+  }
+
+  protected cancelRequest(
+    response: DebugProtocol.CancelResponse,
+    args: DebugProtocol.CancelArguments,
+    _request?: DebugProtocol.Request,
+  ): void {
+    if (args.progressId !== undefined) {
+      if (this.queryRun?.id === args.progressId) {
+        this.cancelEvaluation();
+      }
+    }
+
+    this.sendResponse(response);
+  }
+
+  protected threadsRequest(
+    response: DebugProtocol.ThreadsResponse,
+    request?: DebugProtocol.Request,
+  ): void {
+    response.body = response.body ?? {};
+    response.body.threads = [
+      {
+        id: 1,
+        name: "Evaluation thread",
+      },
+    ];
+
+    super.threadsRequest(response, request);
+  }
+
+  protected stackTraceRequest(
+    response: DebugProtocol.StackTraceResponse,
+    _args: DebugProtocol.StackTraceArguments,
+    _request?: DebugProtocol.Request,
+  ): void {
+    response.body = response.body ?? {};
+    response.body.stackFrames = [];
+
+    super.stackTraceRequest(response, _args, _request);
+  }
+
+  private async launch(
+    response: DebugProtocol.LaunchResponse,
+    args: CodeQLDebugProtocol.LaunchRequestArguments,
+  ): Promise<void> {
+    response.body = response.body ?? {};
+
+    this.args = args;
+
+    void this.evaluate(response);
+  }
+
+  private createLogger(): BaseLogger {
+    return {
+      log: async (message: string, _options: LogOptions): Promise<void> => {
+        this.sendEvent(new OutputEvent(message, "console"));
+      },
+    };
+  }
+
+  private async evaluate(
+    response: DebugProtocol.LaunchResponse,
+  ): Promise<void> {
+    // Send the response immediately. We'll send a "stopped" message when the evaluation is complete.
+    this.sendResponse(response);
+
+    const args = this.args!;
+
+    this.tokenSource = new CancellationTokenSource();
```

Comment:
That's true today. It's possible for a single debug session to support multiple running processes, but there's no motivation to do that now since it's easy enough to let VS Code launch a separate session for each query. If/when we move the debug adapter into the CLI, we'd probably want to implement the multi-launch support so that we can just have one instance of the adapter, running in the query server process.